File: Java/Collatz.java

public class Collatz {

	/**
	 * Función principal.
	 * Comprueba los argumentos y llama a la función para imprimir la serie
	 *
	 * @param args
	 * 		Argumentos pasados por la línea de comandos.
	 */
	public static void main (String args []) {

		long n = 0;

		if (args.length != 1) {

			System.out.println ("Error -> Llamada correcta: "
					+ "\n\tjava Collatz <número_inicial>\n");
			System.exit (-1);
		}

		try {
			n = Long.parseLong (args [0]);
		} catch (NumberFormatException nfe) {

			System.out.printf ("Error: '%s' no es un número válido.", args [0]);
			System.exit (-2);
		}

		System.out.println ("\nNúmero inicial: " + n);
		imprimir_serie (n);
		System.out.println ("---- FIN DE LA SERIE ----\n");
	}

	/**
	 * Imprime la serie desde el número indicado.
	 *
	 * @param n
	 * 		Número inicial de la serie.
	 */
	public static void imprimir_serie (long n) {

		for (int paso = 1; n > 1; paso++) {

			n = ((n % 2) == 0)? (n / 2) : ((3 * n) + 1);
			System.out.printf ("Paso %5d: %10d\n", paso, n);
		}
	}
}

File: Scala/Collatz.scala

/**
 * Objeto en el que se implementan todas las funciones, que se llamarán desde main
 */
object Collatz {

  def main (args: Array [String]) = {

    val n = comprobar_arg (args)

    if (n > 1) {

      println ("\nNúmero inicial: " + n + "\n")
      collatz (n)
      println ("---- FIN DE LA SERIE ----")
    }
  }

  /**
   * Calcula el siguiente término de la serie de Collatz y lo imprime.
   * Si es el último, sale.
   *
   * @param n
   *           Número anterior en la serie.
   * @param paso
   *           Paso actual en la recursividad.
   */
  def collatz (n: Long, paso: Int = 0): Unit = {

    println ("Paso %5d: %10d".format (paso, n))

    if (n > 1) {

      if ((n % 2) == 0) {

        collatz (n / 2, paso + 1);
      } else {

        collatz (3 * n + 1, paso + 1)
      }
    }

  }

  /**
   * Comprueba que esté el argumento necesario
   *
   * @param args
   *        Array con los argumentos pasados por línea de comandos
   *
   * @return
   *        El número cuya serie se debe calcular, o
   *        -1 si no hay un argumento válido
   */
  def comprobar_arg (args: Array [String]): Long = {

    if (args.length != 1) {

      println ("Error -> Llamada correcta: \n\tscala Collatz <número_inicial>");
      -1
    } else {

      val n = args (0)
      if (esNum (n)) n.toLong else -1
    }
  }

  /**
   * Comprueba si la cadena es un número.
   *
   * @param x
   *        Cadena que se desea comprobar
   *
   * @return
   *        true si todos los elementos de la cadena son dígitos; o false si no.
   */
  def esNum (x: String): Boolean = !x.isEmpty && x.forall (Character.isDigit);

}
